### PR TITLE
feat(jsonrpc): add resource restrict for jsonrpc

### LINKED ISSUE #6632: [Feature] Introduce resource limits for JSON-RPC (batch size, response size, address size, timeout)
## Problem Statement
The current JSON-RPC service in java-tron lacks unified constraints on both request scale and response payload size. While this “unbounded” design works in simple scenarios, it introduces significant stability risks when the node is exposed to public traffic or operates under high concurrency.

Specifically:

- For batch array requests, it may contain a large array of sub-requests. A single request can occupy processing threads for an extended period, degrading overall throughput.
- Certain APIs (e.g., `eth_getLogs`, `eth_getFilterChanges`) may return extremely large result sets. Without response size limits, this can lead to excessive memory consumption and even trigger OOM conditions.
- Some requests may become long-running due to I/O latency, holding node resources for prolonged periods.
- Log-related APIs (`eth_getLogs`, `eth_newFilter`) currently allow an unbounded number of address filters, significantly increasing query complexity and directly impacting node performance.

These issues are further amplified when handling untrusted external requests. Therefore, it is necessary to introduce fundamental constraint mechanisms at the JSON-RPC layer.

## Proposed Solution

The implementation prioritizes low-overhead approaches focusing on early validation and resource control.

- For all JSON-RPC requests, if its json type is array, the length should be validated during the parsing phase to enable early rejection, avoiding unnecessary downstream processing.
- For `eth_getLogs` and `eth_newFilter`, the address array should be validated before entering query execution, preventing expensive database or index scans.
- Response size control should not rely on constructing the full payload in memory before validation. Instead, size tracking should be performed incrementally during serialization, using streaming or chunked output to enforce limits and reduce memory pressure.
- Request execution time should be controlled via asynchronous execution. Requests exceeding a configured timeout should be interrupted.

Error handling should remain consistent with the JSON-RPC specification and align with Ethereum conventions for better client interoperability. Different constraint violations should map to clear error codes:

- Address limit exceeded → -32602 (JsonRpcInvalidParamsException)
- Batch size exceeded → -32005
- Response too large → -32003
- Request timeout → -32002

### Specification

Introduce a set of configurable limits for the JSON-RPC service to constrain request size and response payload, with validation performed as early as possible in the request lifecycle to improve system stability and predictability.

The following configuration items are proposed:

1. `node.jsonrpc.maxBatchSize`: Limits the maximum number of sub-requests in a batch JSON-RPC call. Default: 1. Requests exceeding this limit should be rejected immediately after parsing to avoid further processing. Note: values greater than 1 may bypass existing per-method QPS controls.

2.  `node.jsonrpc.maxAddressSize`: Limits the number of address filters in log-related APIs (`eth_getLogs`, `eth_newFilter`). Default: 1000. Validation must occur before query execution to avoid high-cost operations. This follows the same design pattern as the existing node.jsonrpc.maxSubTopics.

3. `node.jsonrpc.maxResponseSize`: Limits the maximum size (in bytes) of a JSON-RPC response. Default: 25MB. The size should be tracked incrementally during serialization. Once the limit is exceeded, the response should be terminated immediately instead of continuing to build the full payload.

4. `node.jsonrpc.maxRequestTimeout`: Limits the execution time of a JSON-RPC request. Default: 30s. Requests exceeding the timeout should be interrupted and return an error.

All configurations should be integrated into the existing configuration system and follow the current JSON-RPC configuration conventions.

**Configuration Changes** 

Add the following configuration items:

- node.jsonrpc.maxBatchSize
- node.jsonrpc.maxResponseSize
- node.jsonrpc.maxAddressSize
- node.jsonrpc.maxRequestTimeout

## Scope of Impact
- API/RPC

**Breaking Changes**
These constraints effectively reduce the impact of oversized or abnormal requests on node resources, mitigate memory risks, and improve stability under high-load conditions. They also enhance the predictability of the JSON-RPC service, making it more suitable for production environments exposed to external users.

**Backward Compatibility**
There are no changes to API fields, so protocol-level compatibility is preserved. However, additional constraints on request inputs and response outputs may require minor client-side adaptations.

## Implementation

**Do you have ideas regarding the implementation?**
Yes

**Are you willing to implement this feature?**
Yes

**Estimated Complexity**
Medium (moderate changes)

## Testing Strategy

Focus primarily on boundary conditions, ensuring that exceptions are triggered as early as possible in the request lifecycle.

**Test Scenarios**

Since request timeout enforcement introduces asynchronous control and may shift execution toward a more serialized model, there could be an impact on performance. Changes in QPS should be evaluated and benchmarked accordingly.

### PR DESCRIPTION
**Pull request by 317787106**
**What does this PR do?**

Adds configurable resource limits to the JSON-RPC endpoint to prevent memory exhaustion and abuse from oversized requests or responses. Closes https://github.com/tronprotocol/java-tron/issues/6632

**Changes:**

1. **Batch size limit (`node.jsonrpc.maxBatchSize`, default: 100)**
   - Validates the array length of batch JSON-RPC requests before dispatching.
   - Requests exceeding the limit are rejected with error code `-32005` (exceed limit).
   - The check is skipped when `maxBatchSize ≤ 0` (no limit).

2. **Response size limit (`node.jsonrpc.maxResponseSize`, default: 25 MB)**
   - Introduces `BufferedResponseWrapper`: intercepts `getOutputStream()` and `getWriter()` writes into an in-memory buffer. When a write would exceed the configured limit, it sets an `overflow` flag and resets the buffer instead of continuing to accumulate bytes, bounding worst-case memory usage to at most `maxResponseSize`.
   - Introduces `CachedBodyRequestWrapper`: replays the pre-read request body via both `getInputStream()` and `getReader()`, so the body can be inspected before being forwarded to `JsonRpcServer`.
   - After the handler returns, the servlet checks `isOverflow()` and — if set — discards the partial buffer and returns error code `-32003` (response too large).

3. **Address list limit (`node.jsonrpc.maxAddressSize`, default: 1000)**
   - In `LogFilter`, validates the `address` array length in `eth_getLogs` / `eth_newFilter` requests.
   - Requests exceeding the limit are rejected with `JsonRpcInvalidParamsException`.

4. **Structured JSON-RPC error responses**
   - `writeJsonRpcError` uses `ObjectMapper` to build error responses safely, avoiding JSON injection from error messages.
   - Error codes follow the JSON-RPC 2.0 spec: `-32700` parse error, `-32005` exceed limit, `-32003` response too large.

**Why are these changes required?**

- Without limits, a client can send an arbitrarily large batch, trigger an expensive query with many addresses, or force the node to serialize a massive response — all of which cause unbounded memory growth.
- The response buffer caps worst-case allocation to `maxResponseSize` and fails fast rather than buffering the entire response before checking.

**Configuration**

```hocon
node {
  jsonrpc {
    # Max JSON-RPC batch array size; 0 = no limit
    maxBatchSize = 100
    # Max response body in bytes (default 25 MB)
    maxResponseSize = 26214400
    # Max address entries in eth_getLogs / eth_newFilter
    maxAddressSize = 1000
  }
}
```

**This PR has been tested by:**
- Unit tests (`BufferedResponseWrapperTest`)
- Manual testing

### CODE REVIEW

File: framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java [hunk 1]
```diff
@@ -66,6 +85,68 @@ public Integer getJsonRpcCode(int httpStatusCode) {
...
+    byte[] body;
+    JsonNode rootNode;
+    try {
+      body = readBody(req.getInputStream());
+      rootNode = MAPPER.readTree(body);
+    } catch (IOException e) {
+      writeJsonRpcError(resp, JsonRpcError.PARSE_ERROR, "Parse error", null);
+      return;
+    }
+    if (rootNode.isArray() && rootNode.size() > parameter.getJsonRpcMaxBatchSize()) {
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java]: [MUST] maxBatchSize=0 does not honor the "0 means no limit" contract

The batch-size guard at line 99 reads `rootNode.size() > parameter.getJsonRpcMaxBatchSize()` without a `> 0` precondition. The PR description and the sibling keys `maxResponseSize` (BufferedResponseWrapper.checkLimit) and `maxAddressSize` (LogFilter) all state that 0 disables the limit; this branch does the opposite — configuring `maxBatchSize = 0` rejects every non-empty batch with `-32005`. That is a silent operator trap.

Suggestion: add a `max > 0 &&` guard so the three limits follow the same semantics.

File: framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java [hunk 2]
```diff
@@ -66,6 +85,68 @@ public Integer getJsonRpcCode(int httpStatusCode) {
...
+        resp, parameter.getJsonRpcMaxResponseSize());
+
+    try {
+      rpcServer.handle(cachedReq, bufferedResp);
+    } catch (JsonRpcResponseTooLargeException e) {
+      JsonNode idNode = (!rootNode.isArray()) ? rootNode.get("id") : null;
+      writeJsonRpcError(resp, JsonRpcError.RESPONSE_TOO_LARGE, e.getMessage(), idNode);
+      return;
+    } catch (Exception e) {
+      throw new IOException("RPC execution failed", e);
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java:112]: [SHOULD] Wrapping every handler exception as IOException escapes as HTTP 500

`catch (Exception e) throw new IOException("RPC execution failed", e)` rethrows all non-`JsonRpcResponseTooLargeException` runtime errors. The parent `RateLimiterServlet.service` only catches `ServletException | IOException` and re-throws, so the servlet container will emit an HTTP 500 with no JSON-RPC body. jsonrpc4j's ErrorResolver would normally have produced a structured error response on HTTP 200 — that contract is now lost, and web3.js/ethers-style clients will interpret 500 as a transport failure and retry.

Suggestion: either drop this catch (let the original IOException path from `rpcServer.handle` propagate) or log the root cause and emit a `-32603 Internal error` through `writeJsonRpcError` so the HTTP 200 + JSON-RPC error contract is preserved.

File: framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java [hunk 3]
```diff
@@ -66,6 +85,68 @@ public Integer getJsonRpcCode(int httpStatusCode) {
...
   @Override
   protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
-    rpcServer.handle(req, resp);
+    CommonParameter parameter = CommonParameter.getInstance();
+
+    byte[] body;
+    JsonNode rootNode;
+    try {
+      body = readBody(req.getInputStream());
+      rootNode = MAPPER.readTree(body);
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java:93]: [SHOULD] Empty or whitespace-only body makes readTree return null and NPE on the next line

`MAPPER.readTree(body)` returns `null` (not a MissingNode) for a zero-length input, so the subsequent `rootNode.isArray()` at line 99 throws NullPointerException. The NPE is not caught by the `IOException` clause at line 95; it bubbles out as an unchecked exception, gets wrapped by the catch at line 116, and the client sees HTTP 500 instead of the structured `-32700 Parse error` that the parse path was meant to return.

Suggestion: treat `rootNode == null || rootNode.isMissingNode()` as a parse error and return `-32700` via `writeJsonRpcError`.

File: framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java [hunk 4]
```diff
@@ -66,6 +85,68 @@ public Integer getJsonRpcCode(int httpStatusCode) {
...
+    } catch (IOException e) {
+      writeJsonRpcError(resp, JsonRpcError.PARSE_ERROR, "Parse error", null);
+      return;
+    }
+    if (rootNode.isArray() && rootNode.size() > parameter.getJsonRpcMaxBatchSize()) {
+      writeJsonRpcError(resp, JsonRpcError.EXCEED_LIMIT,
+          "Batch size " + rootNode.size() + " exceeds the limit of "
+              + parameter.getJsonRpcMaxBatchSize(), null);
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java]: [SHOULD] Batch-size rejection returns a single error object, not a batch array

A JSON-RPC 2.0 batch request must be answered with a JSON array of responses. When `rootNode.isArray()` is true and the size exceeds `maxBatchSize`, the current code returns a single-object error response, which breaks the batch contract — standard clients (web3j, web3.js, ethers) will fail to parse the response as a batch and either error out or retry the whole batch.

Suggestion: when rejecting an over-size batch, wrap the `-32005` error object in a JSON array of length 1 (id=null), or document this deviation prominently in `config.conf` and the PR description.

File: framework/src/main/java/org/tron/core/services/filter/BufferedResponseWrapper.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  private void checkLimit(int incoming) {
+    if (maxBytes > 0 && buffer.size() + incoming > maxBytes) {
+      throw new JsonRpcResponseTooLargeException(
+          "Response byte size exceeds the limit of " + maxBytes);
+    }
+  }
+
+  @Override
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/filter/BufferedResponseWrapper.java:104]: [SHOULD] getStatus() and setHeader("Content-Length", ...) are not intercepted

Header capture only covers `setStatus`, `setContentType`, `setContentLength(int|long)`. Two gaps:

1. `getStatus()` is not overridden, so callers (logging filter, metrics) read the underlying response status, not the `this.status` field — they will see `SC_OK` until `commitToResponse()` is called.
2. `setHeader("Content-Length", ...)` / `addHeader(...)` pass straight through to the underlying response. jsonrpc4j currently calls `setContentLength`, but any downstream filter or library upgrade using `setHeader` would write a Content-Length to the real response before the size check runs.

Suggestion: override `getStatus()` to return `this.status`, and intercept `setHeader` / `addHeader` for the `Content-Length` name (case-insensitive).

File: common/src/main/java/org/tron/core/exception/jsonrpc/JsonRpcResponseTooLargeException.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package org.tron.core.exception.jsonrpc;
+
+public class JsonRpcResponseTooLargeException extends RuntimeException {
```
Comment (bladehan1) [inline common/src/main/java/org/tron/core/exception/jsonrpc/JsonRpcResponseTooLargeException.java]: [SHOULD] New exception breaks the project's JsonRpcException family

Every other class in `common/src/main/java/org/tron/core/exception/jsonrpc/` extends `JsonRpcException` (which in turn extends `TronException` and reports itself). This new class extends `RuntimeException` instead, so it cannot be wired into `@JsonRpcError` annotations or the shared `JsonRpcErrorResolver` if future methods need to surface response-too-large through the normal error path.

Suggestion: extend `JsonRpcException`, or document in a Javadoc why an unchecked non-TronException base was chosen (e.g. to avoid forcing the wrapper signature to `throws`).

File: framework/src/main/java/org/tron/core/services/filter/CachedBodyRequestWrapper.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  private final byte[] body;
+
+  public CachedBodyRequestWrapper(HttpServletRequest request, byte[] body) {
+    super(request);
+    this.body = body;
+  }
+
+  @Override
+  public ServletInputStream getInputStream() {
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/filter/CachedBodyRequestWrapper.java:26]: [SHOULD] getInputStream and getReader are not mutually exclusive

Servlet spec requires that once one of `getInputStream()` / `getReader()` has been used on a request, the other must throw `IllegalStateException`. This wrapper returns a fresh stream/reader from the cached byte array for both methods, allowing arbitrary interleaving. jsonrpc4j only calls one today, so the bug is latent — but any future filter that reads the body through the other accessor would silently double-read.

Suggestion: track which accessor was used first and throw `IllegalStateException` on the second, as the spec requires.

File: common/src/main/java/org/tron/common/parameter/CommonParameter.java [hunk 8]
```diff
@@ -459,6 +459,15 @@ public class CommonParameter {
   @Getter
   @Setter
   public int jsonRpcMaxBlockFilterNum = 50000;
+  @Getter
+  @Setter
+  public int jsonRpcMaxBatchSize = 100;
+  @Getter
+  @Setter
+  public int jsonRpcMaxResponseSize = 25 * 1024 * 1024;
```
Comment (bladehan1) [inline common/src/main/java/org/tron/common/parameter/CommonParameter.java:467]: [SHOULD] Size-limit fields use int; no range validation on config load

`jsonRpcMaxResponseSize` and the peer `maxBytes` parameter in `BufferedResponseWrapper` are `int`, capped at 2 GiB. The current 25 MiB default is fine, but reading via `config.getInt(...)` with no range validation means a negative config value silently becomes a permanent "no limit" state (thanks to the `maxBytes > 0` guard) and a value near `Integer.MAX_VALUE` will put `ByteArrayOutputStream` near its addressable limit.

Suggestion: either switch the field (and wrapper param) to `long` plus validate `>= 0` during `Args.applyConfigParams`, or reject out-of-range values with a clear error at startup.

File: framework/src/main/java/org/tron/core/services/filter/BufferedResponseWrapper.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * <p>If {@code maxBytes > 0}, writes that would push the buffer past {@code maxBytes} throw
+ * {@link JsonRpcResponseTooLargeException} immediately, bounding memory usage to at most
+ * {@code maxBytes} rather than the full response size.
+ *
+ * <p>Header-mutating methods ({@code setStatus}, {@code setContentType}) are buffered here and
+ * only forwarded to the real response via {@link #commitToResponse()}, preventing a timed-out
+ * handler thread from racing with the timeout error writer.
+ */
+public class BufferedResponseWrapper extends HttpServletResponseWrapper {
```
Comment (bladehan1) [inline framework/src/main/java/org/tron/core/services/filter/BufferedResponseWrapper.java:24]: [DISCUSS] Consider Jetty SizeLimitHandler for the response-size cap

Scoped to the response-size concern only — the batch-size and address-size checks must stay at the JSON-RPC layer and are out of scope for this discussion.

The response-size cap this class implements can also be enforced one layer lower. Jetty 9.4 (project is on 9.4.57 per `framework/build.gradle:45`) ships `org.eclipse.jetty.server.handler.SizeLimitHandler`, which wraps the output stream, counts bytes on the fly, and aborts the response when the configured cap is reached. Wrapping the `ServletContextHandler` in `FullNodeJsonRpcHttpService.addServlet` (line 30) with it would:

- Drop the full in-memory buffer. Peak memory per connection falls from ~`maxResponseSize` (the `ByteArrayOutputStream` in `checkLimit`) to the Jetty output buffer (~8 KiB), i.e. true streaming instead of buffer-then-commit.
- Remove this wrapper and its `JsonRpcResponseTooLargeException` plumbing — the same attacker model (oversized response → bounded memory) is still covered.

Tradeoff: SizeLimitHandler aborts with `HttpException.RuntimeException` → HTTP 413 plain-text, not a JSON-RPC `-32003` structured error. Clients like web3j / ethers would see a transport-level failure rather than a parseable JSON-RPC error object. If that structured-error UX is a product requirement (for example so clients can auto-fallback to a smaller range), keep the current wrapper. If the cap is intended as a safety net where 413 is acceptable, SizeLimitHandler is simpler and more memory-efficient.

No change requested — just worth deciding explicitly which tradeoff this PR is making.

File: framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java (hunk 4 above)
Comment (317787106) [reply framework/src/main/java/org/tron/core/services/jsonrpc/JsonRpcServlet.java]: If batch request  exceeds the maxBatchSize, it should returns a single-object error response, Suppose if the batch size is over 100000, should I answer with 10000-size array ?
